R/shoji-catalog.R

```diff
@@ -240,28 +240,109 @@ setMethod("emails", "ShojiCatalog", function (x) getIndexSlot(x, "email"))
 #' @export
 as.list.ShojiCatalog <- function (x, ...) lapply(names(index(x)), function (i) x[[i]])
 
-#' Utility to get a more human-readable view of a Shoji Catalog
+#' A utility to retun a dataframe from a ShojiCatalog.
+#'
+#' Some of the attributes of a ShojiCatalog will not naturally fit in
+#' a conventional dataframe. For instance a single variable might have multiple subvariables,
+#' and these subvariables will not fit in a single row of a dataframe. In this case the
+#' list of subvariables are stored in a list-column in the resulting dataframe.
 #'
 #' @param x ShojiCatalog or subclass
 #' @param keys character vector of attribute names from each catalog tuple to
 #' include in the result. Default is TRUE, which means all.
-#' @param rownames See \code{\link[base]{data.frame}}, the \code{row.names}
-#' argument, to which this is passed in \code{data.frame}. The difference here
-#' is that if \code{rownames} is explicitly set as \code{NULL}, the resulting
-#' object will not have row names set. By default, row names will be the URLs
-#' of the catalog tuples.
+#' @param rownames The rownames of the resulting dataframe. If set to NULL rownames
+#' will default to the resulting dataframe will not have row names. By default the row names
+#' will be the URLs of the catalog tuples.
+#' @param list_columns A character vector of the names of the attributes which should be stored
+#' in a list-column.
 #' @param ... additional arguments passed to \code{data.frame}
 #' @return a \code{data.frame} view of the catalog
 #' @export
-catalogToDataFrame <- function (x, keys=TRUE, rownames, ...) {
-    default.rownames <- missing(rownames)
-    if (default.rownames) {
-        rownames <- NULL
-    }
-    out <- data.frame(do.call(rbind, lapply(index(x), function (a) a[keys])),
-        row.names=rownames, ...)
-    if (default.rownames) {
-        rownames(out) <- NULL
+catalogToDataFrame <- function(x, keys=TRUE,
+    rownames = NULL,
+    list_columns = c("subvariables", "subvariables_catalog"),
+    ...) {
+
+    index <- lapply(index(x), function(a) a[keys])
+    ## Return an empty dataframe if the function is called on an empty catalog
+    if (length(index) == 0) {
+        message("Catalog is empty, returning an empty dataframe")
+        return(data.frame())
+    } else {
+        ### The following code is equivalent to out <- purrr::map_df(index, entry_to_df)
+        ################
+        entry_list <- lapply(index, entry_to_df, list_col_names = list_columns)
+        names   <- unique(unlist( lapply(entry_list, names)))
+        out <- data.frame(matrix(nrow = length(entry_list), ncol = length(names)))
+        names(out) <- names
+
+        for (i in seq_along(entry_list)) {
+            for (j in  names(entry_list[[i]])) {
+                out[i, j] <- entry_list[[i]][1, j]
+            }
+        }
+        #################
+
+        out <- as.data.frame(out, rownames = rownames, ...)
```

Comment:
I think pre-allocating a dataframe and then filling it is a common pattern in R (it's in the [R inferno](http://www.burns-stat.com/pages/Tutor/R_inferno.pdf) anyway). And in this case I find a nested for loop easier to read and understand than a nested `lapply` with a `do.call` around it. I admit that a big part of this comes from not really liking `do.call` as a pattern because of some of the stuff that Hadley mentions [here](http://rpubs.com/hadley/do-call2).

Initially this block of code used `plyr::bind_rows()` which is both much simpler and much more performant than any of the other options I'm aware of. I was told that performance is not an issue for this function because we're always going to be waiting for the API call, and that we didn't want to bring in the dependency.